src/irmin/inode.ml

```diff
@@ -0,0 +1,983 @@
+include Inode_intf
+
+let src = Logs.Src.create "irmin.i" ~doc:"inodes for irmin backend"
+
+module Log = (val Logs.src_log src : Logs.LOG)
+
+let rec drop n (l : 'a Seq.t) () =
+  match l () with
+  | l' when n = 0 -> l'
+  | Nil -> Nil
+  | Cons (_, l') -> drop (n - 1) l' ()
+
+let take : type a. int -> a Seq.t -> a list =
+  let rec aux acc n (l : a Seq.t) =
+    if n = 0 then acc
+    else
+      match l () with Nil -> acc | Cons (x, l') -> aux (x :: acc) (n - 1) l'
+  in
+  fun n s -> List.rev (aux [] n s)
+
+module Make
+    (Conf : S.INODE_CONF)
+    (H : Hash.S)
+    (Node : Node.S with type hash = H.t) =
+struct
+  let () =
+    if Conf.entries > Conf.stable_hash then
+      invalid_arg "entries should be lower or equal to stable_hash"
+
+  module Node = struct
+    include Node
+    module H = Hash.Typed (H) (Node)
+
+    let hash = H.hash
+  end
+
+  module T = struct
+    type hash = H.t [@@deriving irmin]
+    type step = Node.step [@@deriving irmin]
+    type metadata = Node.metadata [@@deriving irmin]
+
+    let default = Node.default
+
+    type value = Node.value
+
+    let value_t = Node.value_t
+    let pp_hash = Type.(pp hash_t)
+  end
+
+  module StepMap = struct
+    include Map.Make (struct
+      type t = T.step
+
+      let compare = Type.(unstage (compare T.step_t))
+    end)
+
+    let of_list l = List.fold_left (fun acc (k, v) -> add k v acc) empty l
+  end
+
+  (* Binary representation, useful to compute hashes *)
+  module Bin = struct
```

Comment:
Maybe `Pre_hash` instead of `Bin` to make the role clearer?

Reply:
This would be a great change but `Bin` has 2 other use cases:
- An interface for the CA: `module CA = Unsafe_CA (H) (Inter.Bin)`
- The serialisation to repr: `Type.map ~pre_hash Bin.t`

We need to extract those 2 behaviors first